#### Algorithm used: Decision Tree

## Data Cleaning 

In [7]:
import pandas as pd
DT_df =  pd.read_csv(r"C:\Users\15878\DT_Reason_file.csv")
DT_df

C:\Users\15878\AppData\Local\Temp\ipykernel_15784\66031282.py:2: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  DT_df =  pd.read_csv(r"C:\Users\15878\DT_Reason_file.csv")


ID  En_ID  Bg_ID  Plant_ID  Line_ID  Machine_ID date   Shift  \
0       824145      1      4         2        1           1  NaN  Shift3   
1       824166      1      4         2        2           1  NaN  Shift3   
2       824172      1      4         2        3           1  NaN  Shift3   
3       824467      1      4         2        1           1  NaN  Shift3   
4       824536      1      4         2        2           1  NaN  Shift3   
...        ...    ...    ...       ...      ...         ...  ...     ...   
58063  1514330      1      4         2        3           1  NaN  Shift1   
58064  1514335      1      4         2        2           1  NaN  Shift1   
58065  1514336      1      4         2        1           1  NaN  Shift1   
58066  1514278      1      4         2        3           1  NaN  Shift2   
58067  1514282      1      4         2        1           1  NaN  Shift2   

        Hour  Global_id  ...  Error_id  type               Start_time  \
0      03-04        3.0  ...       NaN     3  2022-01-21 02:59:41.290   
1      03-04        NaN  ...       NaN     2  2022-01-21 03:05:41.290   
2      03-04        NaN  ...       NaN     2  2022-01-21 03:05:14.123   
3      03-04        NaN  ...       NaN     2  2022-01-21 03:10:35.367   
4      03-04        NaN  ...       NaN     2  2022-01-21 03:10:53.670   
...      ...        ...  ...       ...   ...                      ...   
58063  14-15        1.0  ...       NaN     3                      NaN   
58064  14-15        NaN  ...       NaN     3                      NaN   
58065  14-15        NaN  ...       NaN     3                      NaN   
58066  09-10        1.0  ...       NaN     3                      NaN   
58067  09-10        NaN  ...       NaN     3                      NaN   

                     Stop_time               Shift_Hour down_time  Reason  \
0      2022-01-21 03:05:33.093                      NaN     351.0  9110.0   
1      2022-01-21 03:05:52.630  2022-01-21 22:30:09.047      11.0     0.0   
2      2022-01-21 03:05:57.693  2022-01-21 22:30:09.047      43.0     0.0   
3      2022-01-21 03:10:40.337  2022-01-21 22:30:09.047       5.0     0.0   
4      2022-01-21 03:11:46.827  2022-01-21 22:30:09.047      53.0     0.0   
...                        ...                      ...       ...     ...   
58063                      NaN                      NaN    1542.0  3020.0   
58064                      NaN                      NaN    1542.0     0.0   
58065                      NaN                      NaN    1543.0     0.0   
58066                      NaN                      NaN  232494.0  3020.0   
58067                      NaN                      NaN  232491.0     0.0   

      Priority  Breakdown_Reason  Global_reason  
0          NaN  Want of Material       Material  
1          NaN        Minor Loss     Minor Loss  
2          NaN        Minor Loss     Minor Loss  
3          NaN        Minor Loss     Minor Loss  
4          NaN        Minor Loss     Minor Loss  
...        ...               ...            ...  
58063      NaN          Accident        Machine  
58064      NaN  Shift End Closer            Man  
58065      NaN  Shift End Closer            Man  
58066      NaN          Accident        Machine  
58067      NaN  Shift End Closer            Man  

[58068 rows x 23 columns]

In [8]:
DT_df.dropna()

Empty DataFrame
Columns: [ID, En_ID, Bg_ID, Plant_ID, Line_ID, Machine_ID, date, Shift, Hour, Global_id, Local_id, tpm_id, reason_ID, Error_id, type, Start_time, Stop_time, Shift_Hour, down_time, Reason, Priority, Breakdown_Reason, Global_reason]
Index: []

[0 rows x 23 columns]

In [9]:
DT_df.fillna(0)

ID  En_ID  Bg_ID  Plant_ID  Line_ID  Machine_ID date   Shift  \
0       824145      1      4         2        1           1    0  Shift3   
1       824166      1      4         2        2           1    0  Shift3   
2       824172      1      4         2        3           1    0  Shift3   
3       824467      1      4         2        1           1    0  Shift3   
4       824536      1      4         2        2           1    0  Shift3   
...        ...    ...    ...       ...      ...         ...  ...     ...   
58063  1514330      1      4         2        3           1    0  Shift1   
58064  1514335      1      4         2        2           1    0  Shift1   
58065  1514336      1      4         2        1           1    0  Shift1   
58066  1514278      1      4         2        3           1    0  Shift2   
58067  1514282      1      4         2        1           1    0  Shift2   

        Hour  Global_id  ...  Error_id  type               Start_time  \
0      03-04        3.0  ...       0.0     3  2022-01-21 02:59:41.290   
1      03-04        0.0  ...       0.0     2  2022-01-21 03:05:41.290   
2      03-04        0.0  ...       0.0     2  2022-01-21 03:05:14.123   
3      03-04        0.0  ...       0.0     2  2022-01-21 03:10:35.367   
4      03-04        0.0  ...       0.0     2  2022-01-21 03:10:53.670   
...      ...        ...  ...       ...   ...                      ...   
58063  14-15        1.0  ...       0.0     3                        0   
58064  14-15        0.0  ...       0.0     3                        0   
58065  14-15        0.0  ...       0.0     3                        0   
58066  09-10        1.0  ...       0.0     3                        0   
58067  09-10        0.0  ...       0.0     3                        0   

                     Stop_time               Shift_Hour down_time  Reason  \
0      2022-01-21 03:05:33.093                        0     351.0  9110.0   
1      2022-01-21 03:05:52.630  2022-01-21 22:30:09.047      11.0     0.0   
2      2022-01-21 03:05:57.693  2022-01-21 22:30:09.047      43.0     0.0   
3      2022-01-21 03:10:40.337  2022-01-21 22:30:09.047       5.0     0.0   
4      2022-01-21 03:11:46.827  2022-01-21 22:30:09.047      53.0     0.0   
...                        ...                      ...       ...     ...   
58063                        0                        0    1542.0  3020.0   
58064                        0                        0    1542.0     0.0   
58065                        0                        0    1543.0     0.0   
58066                        0                        0  232494.0  3020.0   
58067                        0                        0  232491.0     0.0   

      Priority  Breakdown_Reason  Global_reason  
0          0.0  Want of Material       Material  
1          0.0        Minor Loss     Minor Loss  
2          0.0        Minor Loss     Minor Loss  
3          0.0        Minor Loss     Minor Loss  
4          0.0        Minor Loss     Minor Loss  
...        ...               ...            ...  
58063      0.0          Accident        Machine  
58064      0.0  Shift End Closer            Man  
58065      0.0  Shift End Closer            Man  
58066      0.0          Accident        Machine  
58067      0.0  Shift End Closer            Man  

[58068 rows x 23 columns]

In [10]:
DT_df.dtypes

ID                    int64
En_ID                 int64
Bg_ID                 int64
Plant_ID              int64
Line_ID               int64
Machine_ID            int64
date                 object
Shift                object
Hour                 object
Global_id           float64
Local_id            float64
tpm_id              float64
reason_ID           float64
Error_id            float64
type                  int64
Start_time           object
Stop_time            object
Shift_Hour           object
down_time           float64
Reason              float64
Priority            float64
Breakdown_Reason     object
Global_reason        object
dtype: object

### Load the data to classifier algorithm

In [5]:
from sklearn.tree import DecisionTreeClassifier

# Create the decision tree classifier object
clf = DecisionTreeClassifier()

# Split the dataset into input features (X) and target variable (y)
X = DT_df['down_time']
y = DT_df['Breakdown_Reason']

# Train the decision tree classifier
clf.fit(X, y)

# Make predictions on new data
new_data = your_new_data
predictions = clf.predict(new_data)

ValueError: Expected 2D array, got 1D array instead:
array=[3.51000e+02 1.10000e+01 4.30000e+01 ... 1.54300e+03 2.32494e+05
 2.32491e+05].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

## Data Preprocessing

### Convert datas into  array 

In [11]:
DT_dict = DT_df['down_time'].to_dict()
print(DT_dict)

{0: 351.0, 1: 11.0, 2: 43.0, 3: 5.0, 4: 53.0, 5: 12.0, 6: 15.0, 7: 12.0, 8: 10.0, 9: 209.0, 10: 210.0, 11: 131.0, 12: 9.0, 13: 51.0, 14: 125.0, 15: 126.0, 16: 1442.0, 17: 19.0, 18: 110.0, 19: 14.0, 20: 111.0, 21: 50.0, 22: 104.0, 23: 93.0, 24: 213.0, 25: 488.0, 26: 55.0, 27: 816.0, 28: 14.0, 29: 14.0, 30: 29.0, 31: 4.0, 32: 2.0, 33: 11.0, 34: 1.0, 35: 73.0, 36: 179.0, 37: 5.0, 38: 17.0, 39: 43.0, 40: 134.0, 41: 625.0, 42: 1.0, 43: 9.0, 44: 30.0, 45: 314.0, 46: 60.0, 47: 7.0, 48: 46.0, 49: 1671.0, 50: 1647.0, 51: 1043.0, 52: 1278.0, 53: 252.0, 54: 246.0, 55: 26.0, 56: 65.0, 57: 48.0, 58: 18.0, 59: 98.0, 60: 79.0, 61: 253.0, 62: 117.0, 63: 104.0, 64: 85.0, 65: 3.0, 66: 456.0, 67: 148.0, 68: 99.0, 69: 83.0, 70: 35.0, 71: 61.0, 72: 127.0, 73: 175.0, 74: 30.0, 75: 101.0, 76: 71.0, 77: 113.0, 78: 414.0, 79: 0.0, 80: 57.0, 81: 157.0, 82: 68.0, 83: 194.0, 84: 141.0, 85: 82.0, 86: 52.0, 87: 60.0, 88: 105.0, 89: 39.0, 90: 75.0, 91: 600.0, 92: 550.0, 93: 146.0, 94: 27.0, 95: 19.0, 96: 38.0, 97: 6

In [12]:
import numpy as np
downtime_array = np.array(list(DT_dict.values()))
print(downtime_array)

[3.51000e+02 1.10000e+01 4.30000e+01 ... 1.54300e+03 2.32494e+05
 2.32491e+05]


In [13]:
DT_dict1 = DT_df['Breakdown_Reason'].to_dict()
print(DT_dict1)

{0: 'Want of Material', 1: 'Minor Loss', 2: 'Minor Loss', 3: 'Minor Loss', 4: 'Minor Loss', 5: 'Minor Loss', 6: 'Minor Loss', 7: 'Minor Loss', 8: 'Minor Loss', 9: 'Minor Loss', 10: 'Communication Loss', 11: 'Communication Loss', 12: 'Minor Loss', 13: 'Minor Loss', 14: 'Minor Loss', 15: 'Minor Loss', 16: 'Tool Change', 17: 'Minor Loss', 18: 'Minor Loss', 19: 'Minor Loss', 20: 'Minor Loss', 21: 'Minor Loss', 22: 'Minor Loss', 23: 'Minor Loss', 24: 'Tea Break Time', 25: 'Tea Break Time Excess', 26: 'Tea Break Time', 27: 'Tea Break Time Excess', 28: 'Minor Loss', 29: 'Minor Loss', 30: 'Minor Loss', 31: 'Minor Loss', 32: 'Minor Loss', 33: 'Minor Loss', 34: 'Minor Loss', 35: 'Minor Loss', 36: 'Minor Loss', 37: 'Minor Loss', 38: 'Minor Loss', 39: 'Minor Loss', 40: 'Minor Loss', 41: 'Want of Material', 42: 'Minor Loss', 43: 'Minor Loss', 44: 'Minor Loss', 45: 'Want of Material', 46: 'Minor Loss', 47: 'Minor Loss', 48: 'Minor Loss', 49: 'Shift End Closer', 50: 'Shift End Closer', 51: 'Cleaning/

In [14]:
import numpy as np
DTReason_array = np.array(list(DT_dict1.values()))
print(DTReason_array)

['Want of Material' 'Minor Loss' 'Minor Loss' ... 'Shift End Closer'
 'Accident' 'Shift End Closer']


### Onehot encoding to convert categorical variable to numerics

In [15]:
import pandas as pd

# Create a DataFrame with the categorical variable
#df = pd.DataFrame({'down_time': [351.0,11.0,43.0,5.0]})

# Perform one-hot encoding
encoded_df = pd.get_dummies(DT_dict)
encoded_df1 = pd.get_dummies(DT_dict1)

# Convert the encoded DataFrame to an array
Downtime_array = encoded_df.values
DowntimeReason_array1 = encoded_df1.values
print(Downtime_array)
print(DowntimeReason_array1)


[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


### Dimensionality reduction

In [16]:
Downtime_array.ndim

2

In [17]:
DowntimeReason_array1.ndim

2

In [ ]:
n = Downtime_array
n1 = DowntimeReason_array1
n.ndim

In [ ]:
x=n.flatten()
x.ndim

In [ ]:
x.ndim

In [ ]:
X.shape

In [1]:
pd.DataFrame(X)

NameError: name 'pd' is not defined

In [ ]:
pd.DataFrame(y)

In [18]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

# Split the dataset into input features (X) and target variable (y)
X = Downtime_array
y = DowntimeReason_array1

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create the decision tree classifier object
clf = DecisionTreeClassifier()

# Train the decision tree classifier using the training data
clf.fit(X_train, y_train)

# Test the classifier using the testing data
accuracy = clf.score(X_test, y_test)

print("Accuracy:", accuracy)


Accuracy: 0.8229722748407095


#### Great! The accuracy of your decision tree classifier on the testing data is 0.8229722748407095, which indicates the percentage of correctly predicted instances. This metric can be used to evaluate the performance of your classifier. However, it's important to note that accuracy alone might not provide a complete picture, especially if the dataset is imbalanced or there are specific requirements for different classes. In such cases, it can be helpful to consider other evaluation metrics like precision, recall, or F1-score.

### Classification Model Evaluation Metrics

In [ ]:
from sklearn.model_selection import cross_val_score


In [ ]:
clf.score(X_test, y_test)

In [ ]:
# Import the necessary libraries
from sklearn.tree import DecisionTreeClassifier

# Create the decision tree classifier object
clf = DecisionTreeClassifier()

# Fit the classifier to the training data
clf.fit(X_train, y_train)

# Make predictions on new data
y_pred = clf.predict(X_test)
print(y_pred)

In [ ]:
pip install tensorflow

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

# Split the dataset into input features (X) and target variable (y)
X = Downtime_array
y = DowntimeReason_array1

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create the decision tree classifier object
clf = DecisionTreeClassifier()

# Train the decision tree classifier using the training data
clf.fit(X_train, y_train)

# Test the classifier using the testing data
accuracy = clf.score(X_test, y_test)

print("Accuracy:", accuracy)




In [ ]:
# Import the necessary libraries
from keras.models import Sequential
from keras.layers import Dense

# Create a sequential model
model = Sequential()

# Add layers to the model
model.add(Dense(units=10, activation='relu', input_shape=(,)))
model.add(Dense(units=num_classes, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Fit the model to the training data
model.fit(X_train, y_train, epochs=10, batch_size=32)

# Make predictions on new data
y_pred = model.predict_classes(X_test)

In [ ]:
from sklearn.ensemble import IsolationForest

# Create the Isolation Forest model
model = IsolationForest(contamination=0.05)  # Set the contamination parameter

# Fit the model to the data
model.fit(y)

# Predict the anomalies
y_pred = model.predict(y)

# The predicted anomalies will have a value of -1, and the normal instances will have a value of 1
anomaly_indices = np.where(y_pred == -1)[0]
print(anomaly_indices)

In [ ]:
anomaly_indices.count